Fixes #11866: first pass at speeding up dhcp (both isc and ms).

dmitri-d opened the pull request.

dmitri-d committed:
Fixes #11866: first pass at speeding up dhcp (both isc and ms).

dmitri-d commented:
This is the first set of changes to speed up dhcp. This PR replaces all but one (here: https://github.com/theforeman/smart-proxy/pull/306/files#diff-d045c300c8c8c9ae152c6c973bb2276eR31) linear searches to constant-time searches.

ekohl commented on modules/dhcp/providers/server/native_ms.rb:
```diff
@@ -186,7 +207,7 @@ def execute cmd, msg=nil, error_only=false
         end
       rescue TimeoutError
-        raise Proxy::DHCP::Error.new("Netsh did not respond within #{tsecs} seconds")
+        raise Proxy::DHCP::Error.new("Netsh did nota respond within #{tsecs} seconds")
```
Typo introduced here.

domcleal replied:
Fixed this on merge.

domcleal commented:
It's kind of difficult with the diff to see what the refactoring achieves.  Please correct me if necessary, but I understand it to be:

1) Move storage from the subnet to the subnet_service, move all data lookups queries through the server instead of dhcp_api using the subnet directly.
2) Keep interactions with the subnet_service largely in the server base class, extend methods in subclasses to add the "real" create/destroy actions etc.
3) The lifecycle of a server is a single request, and so is the subnet_service - so the cache is only used for one request and doesn't have any invalidation issues etc because of that.

dmitri-d commented:
@domcleal: Correct on all three items. I tried to decouple `Proxy::DHCP::Server`, `Proxy::DHCP::Subnet`, and `Proxy::DHCP::Record` from each other (to make the whole thing testable) and preserve existing behavior.

domcleal commented on modules/dhcp/providers/server/isc.rb:
```diff
@@ -104,38 +91,72 @@ def loadSubnetData subnet
+      load_subnet_data(parse_config_and_leases)
+    end
+
+    def load_subnets
```
This is incredibly similar naming to loadSubnets, can they be substantially different please?  Or incorporate this straight into loadSubnets, or rename loadSubnets to load_subnets.  (It would be nice to remove the camelCase names.)

domcleal commented on modules/dhcp/dhcp_api.rb:
```diff
@@ -16,17 +16,28 @@ class Proxy::DhcpApi < ::Sinatra::Base
                                        :config => Proxy::DhcpPlugin.settings.dhcp_config,
-                                       :leases => Proxy::DhcpPlugin.settings.dhcp_leases)
+                                       :leases => Proxy::DhcpPlugin.settings.dhcp_leases,
+                                       :service => Proxy::DHCP::SubnetService.new(Proxy::MemoryStore.new, Proxy::MemoryStore.new,
```
Could these be optimised so it's not repeated for each provider?  Perhaps by initialising sane defaults inside the Proxy::DHCP::SubnetService.

dmitri-d replied:
Nods, this is pretty unseemly.

domcleal commented on modules/dhcp/subnet_service.rb:
```diff
@@ -0,0 +1,98 @@
+class Proxy::DHCP::SubnetService
+  # rubocop:disable Metrics/ParameterLists
```
missing an :enable?

domcleal commented:
The lifecycle of the store object was initially my biggest concern until I realised that it was only going to last for a single request.  That being the case, are you planning to use it for longer-term caching at any point as it has quite a few flaws if that was the plan?

As it stands, I'm not sure it's worth making alterations after adding/deleting records if it's not actually going to last for long.  I also wonder how much it's needed in other cases.

dmitri-d commented:
Nods, there's no caching happening atm, as all stores live only for the duration of a request. I think the biggest benefit of this PR is simplified testing, as speed improvements are marginal. The most expensive bit -- parsing of leases file is still present. 

I'd like to try replacing all operations on single records (view/add/delete) to omapi calls (in windows netsh should be able to provide similar functionality) -- we'll be trading lease file parsing for network latency, but the latter should be much cheaper for large networks. I think the only place where caching _might_ be useful is free IP address lookups.

I agree re: memory store updates on record additions/deletions, I'll remove those. Lemme know if you have questions/concerns.

dmitri-d commented:
@domcleal: renamed subnet- and record-loading methods in `Proxy::DHCP::Server` and subclasses, cleaned up provider instantiation. Memory store is no longer being updated on add/remove record operations.

shlomizadok commented on modules/dhcp/providers/server/isc.rb:
```diff
@@ -104,38 +96,72 @@ def loadSubnetData subnet
+      ret_val
+    end
+
+    def loadSubnets
```
nitpick: rename this method name? (all over)

dmitri-d replied:
It's a legacy name, I'd like to minimize renames in older parts of the code at this point.

domcleal commented on modules/dhcp/subnet_service.rb:
```diff
@@ -0,0 +1,105 @@
+  # rubocop:enable Metrics/ParameterLists
+
+  def add_subnet(subnet)
+    raise Proxy::DHCP::Error, "Unable to Add Subnet #{subnet}" if find_subnet(subnet.network)
```
Nitpick. lower case "add subnet"

domcleal commented on modules/dhcp/subnet_service.rb:
```diff
@@ -0,0 +1,105 @@
+    @leases_by_mac[subnet_address, record.mac] = record
+  end
+
+  def add_host(subnet_address, record)
```
Could we have some debug or trace level logging in these methods?  I think it might be useful - we used to in add_record.

domcleal commented on test/dhcp/subnet_service_test.rb:
```diff
@@ -0,0 +1,187 @@
+               Proxy::DHCP::Subnet.new("192.168.1.0", "255.255.255.0")]
+    @service.add_subnets(*subnets)
+
+    assert_equal subnets.first, @service.find_subnet("192.168.0.0")
```
This should also be tested for the latter part of the method, to ensure it can find a subnet by an IP within the net, not just the network address.

domcleal commented on modules/dhcp/dhcp_api.rb:
```diff
@@ -92,6 +105,10 @@ def load_subnet
   # create a new record in a network
   post "/:network" do
     begin
+
```
Nitpick, remove blank lines at beginning of the block, ditto below.

domcleal commented:
Sorry for the slow trickle of review comments, I've been sidetracked.  I'd suggest waiting until I set the WoC label before trying to address them - thanks.

dmitri-d commented:
@domcleal: rebased and fixed things you pointed out (good catch re: missing test).

domcleal commented on modules/dhcp/providers/server/isc.rb:
```diff
@@ -25,7 +32,6 @@ def delRecord subnet, record
       omcmd "remove"
       omcmd("disconnect", msg)
-      subnet.delete record
     end
```
missing super() call in delRecord?

domcleal replied:
Ignore, thought I saw it defined in the Server class.

dmitri-d replied:
It used to be, but it was an empty method, didn't think there was any point to keep it around.

domcleal commented:
Okay, so I'm pretty happy with it and it tests well.  I'm not sure that we should use the associated ticket #6916 though, since I don't imagine that this fixes the issue - my understanding is that the native MS provider's as slow as it is because of the enumeration of subnets (which is still present), not the lookups.

dmitri-d commented:
@domcleal: there are still performance-related issues in both MS and ISC providers, this PR only addresses the "low-hanging fruit" -- various linear searches. If you don't think it's a sufficient justification I can open a new ticket instead.

domcleal commented:
@witlessbird if you could, I'd appreciate it.  I'm hesitant to close that one right now.

dmitri-d commented:
@domcleal: done and done: http://projects.theforeman.org/issues/11866

domcleal commented:
Merged as ce9ff4c, thanks @witlessbird.
